### PR TITLE
Merge sharding into master branch

### PR DESCRIPTION
**Pull request by hwwhww**
### What was wrong?
#374

A glimpse of what we need to clean.

#### Cute Animal Picture

put a cute animal picture link inside the parentheses

### CODE REVIEW

File: evm/chains/chain.py [hunk 1]
```diff
@@ -1,405 +1,7 @@
...
-)
-from evm.utils.hexadecimal import (
-    encode_hex,
-)
-from evm.utils.rlp import (
-    ensure_imported_block_unchanged,
-)
-
-
-class Chain(Configurable):
```
Comment (hwwhww) [inline evm/chains/chain.py]: Note to self: make sure that the latest and `Chain` is moved to `BaseChain` completely.

File: evm/rlp/headers.py [hunk 2]
```diff
@@ -148,3 +159,117 @@ def clone(self):
...
+                 shard_id,
+                 expected_period_number,
+                 period_start_prevhash,
+                 parent_hash,
+                 number,
+                 transaction_root=EMPTY_SHA3,
+                 coinbase=ZERO_ADDRESS,
+                 state_root=EMPTY_SHA3,
+                 receipt_root=EMPTY_SHA3,
+                 sig=b""):
```
Comment (hwwhww) [inline evm/rlp/headers.py]: TODO: No `sig` now.

File: evm/utils/db.py [hunk 3]
```diff
@@ -30,3 +41,22 @@ def get_block_header_by_hash(block_hash, db):
     Returns the header for the parent block.
     """
     return db.get_block_header_by_hash(block_hash)
+
+
+def get_empty_root_hash(db):
+    # TODO: Backport and refactor evm.chains.base.BaseChain and
```
Comment (hwwhww) [inline evm/utils/db.py]: TODO: Remove this comment.

File: evm/vm/forks/sharding/computation.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+)
+from evm.utils.hexadecimal import (
+    encode_hex,
+)
+from evm.validation import (
+    validate_uint256,
+)
+
+from ..spurious_dragon.computation import (
+    SpuriousDragonComputation,
```
Comment (hwwhww) [inline evm/vm/forks/sharding/computation.py]: TODO: Use `ByzantiumComputation`.

File: evm/exceptions.py [hunk 5]
```diff
@@ -141,3 +149,25 @@ class OutOfBoundsRead(VMError):
...
+
+
+class GasPriceAlreadySet(Exception):
+    """
+    Error raised if trying to reset the already set gas price.
+    """
+    pass
+
+
+class NotTopLevelCall(Exception):
```
Comment (hwwhww) [inline evm/exceptions.py]: @NIC619  ditto

File: evm/exceptions.py [hunk 6]
```diff
@@ -141,3 +149,25 @@ class OutOfBoundsRead(VMError):
...
+
+class UnannouncedStateAccess(VMError):
+    """
+    Error raised if a storage slot or account field is accessed that has not
+    been specified in the transaction's read or write list, respectively.
+    """
+    pass
+
+
+class GasPriceAlreadySet(Exception):
```
Comment (hwwhww) [inline evm/exceptions.py]: @NIC619  Should it be a `VMError`?
```python
class GasPriceAlreadySet(Exception):
```

File: evm/vm_state.py [hunk 7]
```diff
@@ -97,14 +101,13 @@ def state_db(self, read_only=False):
         elif self.state_root != state.root_hash:
             self.set_state_root(state.root_hash)
 
-        self.access_logs.reads.update(state.db.access_logs.reads)
-        self.access_logs.writes.update(state.db.access_logs.writes)
+            self.access_logs.reads.update(state.db.access_logs.reads)
```
Comment (hwwhww) [inline evm/vm_state.py]: Now I think `self.access_logs.reads.update(state.db.access_logs.writes)` should be in `elif self.state_root != state.root_hash` section while `self.access_logs.writes.update(state.db.access_logs.reads)` isn't.

File: evm/exceptions.py (hunk 6 above)
Comment (pipermerriam) [reply evm/exceptions.py]: I think it's intentionally not a VM error to ensure that mis-use of this API results in a hard failure, requiring EVM code that uses this API to correctly handle all of the cases (such as the price already being set).

File: evm/vm_state.py (hunk 7 above)
Comment (NIC619) [reply evm/vm_state.py]: Seems like a typo? self.access_logs.**reads**.update(state.db.access_logs.**writes**)
I suppose `self.access_logs.reads.update(state.db.access_logs.reads)` should be updated regardless of whether it's read or write?

File: evm/computation.py [hunk 8]
```diff
@@ -303,6 +304,11 @@ def get_gas_remaining(self):
         else:
             return self.gas_meter.gas_remaining
 
+    @contextmanager
+    def state_db(self, read_only=False):
```
Comment (NIC619) [inline evm/computation.py]: Note: #405 split `state_db` into `read_only_state_db` and `mutable_state_db`.
I can wait for this one to get merged first and make consequent updates.

File: evm/rlp/headers.py (hunk 2 above)
Comment (hwwhww) [reply evm/rlp/headers.py]: ^^^^ After merging, use #420 to fix it.

File: evm/exceptions.py (hunk 5 above)
Comment (pipermerriam) [reply evm/exceptions.py]: The way it's used, I don't believe so.  This isn't something that halts VM execution, but rather to signal to the opcode logic that it's being called incorrectly and to push `0` onto the stack.

File: tox.ini [hunk 9]
```diff
@@ -1,6 +1,7 @@
 [tox]
 envlist=
     py{35,36}-{core,p2p,database,transactions,vm,native-blockchain}
+    py{35,36}-{auxiliary, sharding}
```
Comment (cburgdorf) [inline tox.ini]: nitpick: I think this should go into the line above, no?

File: evm/logic/call.py [hunk 10]
```diff
@@ -58,8 +58,9 @@ def __call__(self, computation):
         computation.gas_meter.consume_gas(child_msg_gas_fee, reason=self.mnemonic)
 
         # Pre-call checks
-        state_db = computation.vm_state.read_only_state_db
-        sender_balance = state_db.get_balance(computation.msg.storage_address)
+        with computation.state_db(read_only=True) as state_db:
+            sender_balance = state_db.get_balance(computation.msg.storage_address)
+
```
Comment (NIC619) [inline evm/logic/call.py]: TODO: After this is merged, I will make a PR to remove `Message.access_list` and `Computation.state_db` which should only apply to `ShardingVM` and another PR to convert this `state_db(read_only=True)`/`state_db()` back to `read_only_state_db`/`mutable_state_db`.

File: tox.ini (hunk 9 above)
Comment (hwwhww) [reply tox.ini]: yeah, I think so.  👍
(we were trying to isolate sharding stuff, but they haven't been separated in `core` anyway)
I will modify it after merging.

**Pull request merged**